### PR TITLE
Ldy

### PR DESCRIPTION
**Pull request by danyang123**
"plot_bar_SEU" and "plot_bar_SE"  are used to show the pre-training and post-training changes of PITime and PIturn for all fish with the DataQuality above 0.95.
" S " means " Self-control ", "E" means " Experiment " and "U" means " Unpaired control".

### CODE REVIEW

File: @ABLITZER/generate_output.m [hunk 1]
```diff
@@ -0,0 +1,40 @@
+% Generate an output struct for checking all results in the same place
+function output = generate_output(a)
```
Comment (young24) [inline @ABLITZER/generate_output.m]: Can we make this function as a method of the ABLITZER class?

File: @ABLITZER/plotOntogenyByPI.m [hunk 2]
```diff
@@ -1,15 +1,16 @@
 % plot performance versus fish age
 % Here, we use the substraction of PI in training with PI in baseline as
 % the characteristic PI for fish at the same age
+% CAUTION: NOT CORRECT
 function  plotOntogenyByPI(obj,metricType)
```
Comment (young24) [inline @ABLITZER/plotOntogenyByPI.m]: We may need to rewrite this function so that it would generate figures just as Fig. 4 in the manuscript.

File: @ABLITZER/processOneDayYamls.m [hunk 3]
```diff
@@ -68,26 +68,26 @@
     'PITurn_Baseline',[],'PITurn_Training',[],'PITurn_Test',[],...
     'NumShock',[],'PIShock',[]);
     for i = 1:length(obj.FishStack)% number of fish in the fishStack
```
Comment (young24) [inline @ABLITZER/processOneDayYamls.m]: It might be better to use the method "generate_output" to replace the following lines.

File: @ABLITZER/visualizeLearningResponses.m [hunk 4]
```diff
@@ -1,110 +1,110 @@
...
+    text(0.1,0.9,str);
+
+    % Add formatted T-test and significance bar, minimum principle
+    sigSyms = ["n.s.",... % P > 0.05
+                "*",...   % P < 0.05
+                "**",...  % P < 0.01
+                "***",... % P < 0.001
+                "****"];  % P < 0.0001 % occassionally use
+
+   % control group: compare mean of PIs between Baseline(C) and Test(C)
```
Comment (young24) [inline @ABLITZER/visualizeLearningResponses.m]: Considering replace these lines with the method "significanceTest"? However, be careful about the difference between the paired t-test and the unpaired t-test. The number of input arguments may vary.

File: @ABLITZER/visualizeLearningResponses.m [hunk 5]
```diff
@@ -1,110 +1,110 @@
...
+    end
+
+    if isempty(PIs)
+        fprintf('No valid data to plot\n');
+        return
+    end
+
+    figure;
+
+    colors = 0.8*ones(6,3);%[0.5,0,0;0,0.5,0;0,0,0.5;0.5,0.5,0;0.5,0,0.5;0,0.5,0.5];
```
Comment (young24) [inline @ABLITZER/visualizeLearningResponses.m:58]: It might be better to replace this part of plotting code with a sub-function since it's more modular.

File: @ABLITZER/yaml2matlab.m [hunk 6]
```diff
@@ -154,13 +154,7 @@ function yaml2matlab(obj,oldFlag,pathName,fileName)
...
     end
     F(i).yDivide = yDivide(i);
-    if contains(task,'control','IgnoreCase',true)
-        F(i).ExpType = "Control Group";
-    elseif contains(task,'exp','IgnoreCase',true)
-        F(i).ExpType = "Exp Group";
-    else
-        F(i).ExpType = "Unrecognized";
-    end
+    F=judge_ExpType(task,F,i);
```
Comment (young24) [inline @ABLITZER/yaml2matlab.m:157]: Good job here. 👍

File: @FISHDATA/calcReactTimeByTrials.m [hunk 7]
```diff
@@ -10,4 +11,46 @@
...
+function TrMat = getTrIndices(obj)
+
+% 1st col: index of begin, 2nd col: index of end, 3rd col: expPhase
+TrMat = [];
+
+expPhase = cat(1,obj.Frames.ExpPhase);
+pIdx = cat(1,obj.Frames.PatternIdx);
+
+% for baseline
+idxPhase = find(expPhase == 0);
```
Comment (young24) [inline @FISHDATA/calcReactTimeByTrials.m]: Would be better if the sub-function used for these repetitive lines.

File: @FISHDATA/measureExtinction.m [hunk 8]
```diff
@@ -5,17 +5,25 @@
...
+
+% TODO: rewrite this function to find the extinction point
+=======
+>>>>>>> 850eab9a93378e9610381f511732d13c006d2b1f
+function extTime = measureExtinction(obj,plotFlag)
     testInterval = 120; % seconds
     frameRate = obj.FrameRate;
     windowWidth = testInterval * frameRate;
     PIthre = obj.Res.PItime(1).PIfish;
     pMov = calc_moving_PI(obj,windowWidth);
```
Comment (young24) [inline @FISHDATA/measureExtinction.m]: Do we need this function to get the extinction point?

File: @FISHDATA/personalAnalysis.m [hunk 9]
```diff
@@ -40,7 +41,7 @@
             titleStr = "Number of crossing mid-line";
     end
     title(titleStr);
-    
+
```
Comment (young24) [inline @FISHDATA/personalAnalysis.m]: Better to use the significanceTest to replace this code block.

File: @FISHDATA/personalAnalysis.m [hunk 10]
```diff
@@ -1,7 +1,8 @@
 % obj personal analysis to see whether obj
-% learned the task and to measure performance 
+% learned the task and to measure performance
 % from several perspectives
 % idxPostTr: indices of post trials for test
+% TODO: rewrite this function to fit the current trial definition
 function TrRes = personalAnalysis(obj,idxPostTr)
 
 TrRes = zeros(24,5);
```
Comment (young24) [inline @FISHDATA/personalAnalysis.m:8]: Better to write/Use a function to replace all lines for plotting though there are just several lines.

File: @FISHDATA/ratePerformanceByTrials.m [hunk 11]
```diff
@@ -1,29 +1,29 @@
 % Rate performance by trials, every 2 mins is counted as a trial
 % TODO: decouple metrics from TrRes
 function  TrRes = ratePerformanceByTrials(obj,metric,plotFlag)
-
+% obj is the Data of one fish
```
Comment (young24) [inline @FISHDATA/ratePerformanceByTrials.m:4]: Let's work on the modularization together for this function.

File: @FISHDATA/sayIfLearned.m [hunk 12]
```diff
@@ -1,8 +1,9 @@
-% Measure whether fish learned or not 
+% Measure whether fish learned or not
 % By dividing the entire process into trials (2 mins as a trial)
 % measure fish performance in each trial, to decide whether fish learned or
-% not and measure the memory extinction  
+% not and measure the memory extinction
 % TODO: Modularize code
+% Caution: not correct, do not use
```
Comment (young24) [inline @FISHDATA/sayIfLearned.m:6]: Good job here.

File: Utility functions/plot_bar_SE.m [hunk 13]
```diff
@@ -0,0 +1,36 @@
+function [output_OLcontrol,output_OLexp,p] = plot_bar_SE(a)
```
Comment (young24) [inline Utility functions/plot_bar_SE.m]: Great work here. However, it would be better to use sub-functions to replace the repetitive lines.

File: Utility functions/plot_bar_SEU.m [hunk 14]
```diff
@@ -0,0 +1,40 @@
+function [output_OLcontrol,output_OLexp,output_Unpaired,p] = plot_bar_SEU(a)
```
Comment (young24) [inline Utility functions/plot_bar_SEU.m]: Great, however, modularize it.

File: Utility functions/plotpairbar.m [hunk 15]
```diff
@@ -0,0 +1,29 @@
+function plotpairbar(y1,y2,y_std,XTickLabel,YLabel)
```
Comment (young24) [inline Utility functions/plotpairbar.m:1]: What's the difference between this function and the "plot_bar_SE"?

**Review by young24 — COMMENTED** [review]
Great job, generally, and it would be better if more sub-functions used.

File: @ABLITZER/yaml2matlab.m [hunk 16]
```diff
@@ -306,7 +306,7 @@ function yaml2matlab(obj,oldFlag,pathName,fileName)
     end
 end
 
-frames(idxFrame:end,:) = []; % remove redundant frames
+frames(idxFrame+1:end,:) = []; % remove redundant frames
```
Comment (young24) [inline @ABLITZER/yaml2matlab.m:309]: Great.

**Comment by danyang123** [comment]
@young24  I have modified some functions again. Can you check what else needs to be modified?

**Comment by young24** [comment]
Sure, will do tomorrow

File: @ABLITZER/yaml2matlab.m (hunk 6 above)
Comment (young24) [inline @ABLITZER/yaml2matlab.m:157]: Great design here. 👍

File: @FISHDATA/personalAnalysis.m [hunk 17]
```diff
@@ -12,6 +13,10 @@
 TrRes(:,5) = obj.ratePerformanceByTrials('crossMidline',false);
 
 % Plot results
+plotfigure(TrRes,idxPostTr);
```
Comment (young24) [inline @FISHDATA/personalAnalysis.m:16]: Good. We might need to use this function in the function "visualizeLearningResponses".

File: @FISHDATA/sayExtTimeandIfLearned.m [hunk 18]
```diff
@@ -0,0 +1,28 @@
...
+% As I know, there are at least 2 kinds of memory: retention and extinction
+% 1. when the CS pattern continues to show on but without electric shock
+% the fish may un-learn the association.
+% 2. after the fish learned, put them in normal raising house for a while,
+% then test whether they will show escaping responses to CS pattern.
+% In this experiment, we measured the first kind of memory
+
+% TODO: rewrite this function to find the extinction point
+
+function [extTime,num_trail,Test_trail,meanBT] = sayExtTimeandIfLearned(obj)
```
Comment (young24) [inline @FISHDATA/sayExtTimeandIfLearned.m:11]: sayExtTimeandIfLearend -> sayExtTime**A**ndIfLearend

File: Utility functions/plotBarIflearned.m [hunk 19]
```diff
@@ -0,0 +1,49 @@
+function [p,Ls,nLs,All] = plotBarIflearned(Ls,nLs)
```
Comment (young24) [inline Utility functions/plotBarIflearned.m:1]: the name of this function is not clear to me. Would the "plotLearnersWithNonlearners" better?

File: Utility functions/plotBarSE.m [hunk 20]
```diff
@@ -0,0 +1,36 @@
+function [OLcontrol,OLexp,p] = plotBarSE(a)
```
Comment (young24) [inline Utility functions/plotBarSE.m:1]: We need to rename this function to make it clearer.

File: Utility functions/plotBarSEU.m [hunk 21]
```diff
@@ -0,0 +1,47 @@
+function [OLcontrol,OLexp,Unpaired,p] = plotBarSEU(a)
```
Comment (young24) [inline Utility functions/plotBarSEU.m:1]: We need to rename this function to make it clearer.

File: Utility functions/plotIflearnersPITimeByTrail.m [hunk 22]
```diff
@@ -0,0 +1,32 @@
+function [num_Ls,num_nLs]=plotIflearnersPITimeByTrail(Ls,nLs)
```
Comment (young24) [inline Utility functions/plotIflearnersPITimeByTrail.m:1]: would "cmpLearnersWithNonlearnersByTrs" better?

**Review by young24 — APPROVED** [review]
I think we can merge your branch first. Then let's revise the remain issues together on file by file. Next time, when you finish one file, you can open a pull request, which would make the reviewing easier for me.
